internal/migration/migration.go

```diff
@@ -0,0 +1,430 @@
+	startedAt := time.Now()
+
+	if err := shouldMigrate(logger, sourceBeaconPath, beaconName, destination, pgDSN); err != nil {
+		logger.Warnw("decided storage format migration is not needed", "err", err)
+		if errors.Is(err, ErrMigrationNotNeeded) {
+			return err
+		}
+
+		return nil
+	}
+
+	bufferSize, err := computerBufferSize(bufferSize, logger, sourceBeaconPath)
+	if err != nil {
+		return err
+	}
+
+	m := migrator{
+		logger:    logger,
+		startedAt: startedAt,
+
+		pgDSN:            pgDSN,
+		bufferSize:       bufferSize,
+		beaconName:       beaconName,
+		sourceBeaconPath: sourceBeaconPath,
+		destination:      destination,
+	}
+
+	return m.doMigrate()
+}
+
+func computerBufferSize(bufferSize int, logger log.Logger, sourceBeaconPath string) (int, error) {
+	if bufferSize < 0 {
+		logger.Infow("buffer size not specified, defaulting to 10000")
+		return DefaultBufferSize, nil
+	}
+
+	if bufferSize == 0 {
+		return automaticBufferSize(logger, sourceBeaconPath)
+	}
+
+	if bufferSize < DefaultBufferSize {
+		logger.Warnw("buffer size seems a bit too small. The migration process might be slow", "bufferSize", bufferSize)
+	}
+
+	//nolint:gomnd // See below
+	if bufferSize > 10_000_000 {
+		//nolint:lll // This line has the right amount of chars
+		logger.Warnw("buffer size seems a bit too large. Make sure your system can allocate enough system memory for this", "bufferSize", bufferSize)
+	}
+	return bufferSize, nil
+}
+
+func automaticBufferSize(logger log.Logger, sourceBeaconPath string) (int, error) {
+	logger.Warnw("buffer size set to 0. Running automatic buffer inference. Make sure you have enough system memory for this")
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return -1, err
+	}
+	defer existingDB.Close()
+
+	var bufferSize = 0
+
+	err = existingDB.View(func(tx *bbolt.Tx) error {
+		bufferSize = tx.Bucket(bucketName).Stats().KeyN
+		return nil
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	logger.Infow("buffer size automatically inferred from existing DB", "bufferSize", bufferSize)
+
+	return bufferSize, nil
+}
+
+func shouldMigrate(logger log.Logger, sourceBeaconPath, beaconName string, destination chain.StorageType, pgDSN string) error {
+	//nolint:exhaustive // We want to explicitly ignore the chain.MemDB backend since there's nothing to migrate there.
+	switch destination {
+	case chain.BoltDB:
+		return shouldMigrateBolt(sourceBeaconPath)
+	case chain.PostgreSQL:
+		return shouldMigratePostgres(logger, beaconName, pgDSN)
+	default:
+		return fmt.Errorf("unknown destination type %q for migration package", destination)
+	}
+}
+
+func shouldMigrateBolt(sourceBeaconPath string) error {
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return err
+	}
+	defer existingDB.Close()
+
+	return existingDB.View(func(tx *bbolt.Tx) error {
+		_, value := tx.Bucket(bucketName).Cursor().First()
+		var b = beacon{}
+		if json.Unmarshal(value, &b) != nil {
```

Comment:
Could this just be replaced with a `nil` check for value? Or are we assuming that if json parsing fails then the binary is at a new version than the tool?

Reply:
The assumption here was that the first/all values are correctly stored in the database.
When trying to unmarshal from the new, raw, format, this will produce an error and we can use that as the flag.
We can come up with a more complex scenario for validating if this is a migrated database or not if desired.